pkg/config/config.go

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"fmt"
+	ioutil "io/ioutil"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"sigs.k8s.io/controller-runtime/pkg/config/v1alpha1"
+)
+
+// ControllerConfiguration defines the functions necessary to parse a config file
+// and to configure the Options struct for the ctrl.Manager
+type ControllerConfiguration interface {
+	runtime.Object
+
+	// GetControllerConfiguration returns the configuration for ctrl.Manager
+	GetControllerConfiguration() v1alpha1.ControllerConfiguration
+}
+
+// FromFile will load the config file and object and populate the fields in the
+// runtime.Object.
+func FromFile(scheme *runtime.Scheme, path string, cfg ControllerConfiguration) (ControllerConfiguration, error) {
```

Comment:
so, I think there's two tweaks we can make that might improve the ergonomics of this a smidge.  The first, discussed above, is to just compose a scheme on the fly when possible (see above).

The second is that we can probably move `scheme`, when desired, to an option, same with cfg.

That way, if you're using the default CR component config, it looks like `.AndFrom(cfg.File("/some path"))`.

Finally, depending on how far we want to go, we can chose a default path, like we do for the `/var/run/kubernetes` stuff.  Not sure if we want to do that, might it's generally a reasonable k8s pattern.

With all three, you end up with:

```go
.AndFrom(cfg.FileOrDie()) // default config, default path /var/run/controller-runtime/cfg.yaml
.AndFrom(cfg.FileOrDie(cfg.AtPath("~/dev/test-config.yaml")) // default config, custom path
.AndFrom(cfg.FileOrDie(cfg.OfKind(&myCustomConfig))) // default path, custom config
```

(that may be a few to many parens, though, but I imagine the first and third forms will be most common)

Aside: an alternate option looks like

```go
.AndFrom(cfg.File{Into: &myCustomConfig})
.AndFrom(&myCustomConfig{
  // filled in
})
```

This is a bit more radical, but it means that ControllerConfiguration becomes:

```
type ControllerConfig interface {
  runtime.Object
  GetControllerConfig() v1alpha1.ControllerConfig
  Finalize() error
}
```

Then, `v1alpha1.ControllerConfig` implements `GetControllerConfig` and `Finalize` by default, so concrete types just work (with `func (ControllerConfig) Finalize() error { return nil }`), and `FromFile` looks like

```go
type File struct {
  AtPath string
  UsingScheme scheme.Scheme
  Into ControllerConfig
}
func (f *File) Finalize() error {
  // do the magic file loady stuff
}
func (f *File) GetControllerConfig() v1alpha1.ControllerConfig {
  return f.Into.GetControllerConfig()
}

// Alternatively, just make GetControllerConfig return an error & use a sync.Once in File,
// but I'm not sure which is move obvious

// do FileOrDie to cause Finalize to panic instead
```